plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt

```diff
@@ -27,49 +33,65 @@ class CodeGenerationState(
-            amazonqRepositorySize = repositorySize,
-        )
-
-        val nextState = PrepareCodeGenerationState(
-            tabID = tabID,
-            approach = approach,
-            config = config,
-            filePaths = codeGenerationResult.newFiles,
-            deletedFiles = codeGenerationResult.deletedFiles,
-            references = codeGenerationResult.references,
-            currentIteration = currentIteration + 1,
-            uploadId = uploadId,
-            messenger = messenger,
-        )
-
-        // It is not needed to interact right away with the PrepareCodeGeneration.
-        // returns therefore a SessionStateInteraction object to be handled by the controller.
-        return SessionStateInteraction(
-            nextState = nextState,
-            interaction = Interaction(content = "", interactionSucceeded = true)
-        )
+        var result: Result = Result.Succeeded
+        var failureReason: String? = null
+        var numberOfReferencesGenerated: Int? = null
+        var numberOfFilesGenerated: Int? = null
+        try {
+            val response = featureDevClientUtil.startTaskAssistCodeGeneration(
+                conversationId = config.conversationId,
+                uploadId = uploadId,
+                message = action.msg
+            )
+
+            messenger.sendAnswerPart(
+                tabId = tabID,
+                message = message("amazonqFeatureDev.code_generation.generating_code")
+            )
+
+            val codeGenerationResult = generateCode(codeGenerationId = response.codeGenerationId())
+            numberOfReferencesGenerated = codeGenerationResult.references.size
+            numberOfFilesGenerated = codeGenerationResult.newFiles.size
+            val nextState = PrepareCodeGenerationState(
+                tabID = tabID,
+                approach = approach,
+                featureDevClientUtil = featureDevClientUtil,
+                config = config,
+                filePaths = codeGenerationResult.newFiles,
+                deletedFiles = codeGenerationResult.deletedFiles,
+                references = codeGenerationResult.references,
+                currentIteration = currentIteration + 1,
+                uploadId = uploadId,
+                messenger = messenger,
+            )
+
+            // It is not needed to interact right away with the PrepareCodeGeneration.
+            // returns therefore a SessionStateInteraction object to be handled by the controller.
+            return SessionStateInteraction(
+                nextState = nextState,
+                interaction = Interaction(content = "", interactionSucceeded = true)
+            )
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Code generation failed: ${e.message}" }
+            result = Result.Failed
+            failureReason = e.message
+            throw e
+        } finally {
+            AmazonqTelemetry.codeGenerationInvoke(
```

Comment:
Do we need to add failure telemetry log here as well, or just adding in API call failures (`FeatureDevClientUtil`) is enough ?

Reply:
I think in this event, we're interested in how long the code gen took and if the codegen failed or not. Not only the invocation of startTaskAssistCodeGeneration API. See the implementation in vscode: https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/amazonqFeatureDev/session/sessionState.ts#L304